cpu/sources/layout7/cpu_oods.move

```diff
@@ -0,0 +1,940 @@
+            // constraints for the trace and composition polynomials, to the fri_queue array.
+            set_el(ctx, *fri_queue + 1, res % prime);
+
+            // print(&(res % prime));
+
+            // Append the friInvPoint of the current query to the fri_queue array.
+            set_el(ctx, *fri_queue + 2, *borrow(batch_inverse_array, *denominators_ptr + 99));
+
+            // Advance denominators_ptr by chunk size (0x20 * (2+N_ROWS_IN_MASK)).
+            *denominators_ptr = *denominators_ptr + 100;
+
+            *fri_queue = *fri_queue + 3;
+        };
+
+        if (*fri_queue >= *fri_queue_end) {
+            *first_invoking = true;
+            *checkpoint = FB_CHECKPOINT1;
+            true
+        } else {
+            false
+        }
+    }
+
+    /*
+          Computes and performs batch inverse on all the denominators required for the out of domain
+          sampling boundary constraints.
+
+          Since the frieval_points are calculated during the computation of the denominators
+          this function also adds those to the batch inverse in prepartion for the fri that follows.
+
+          After this function returns, the batch_inverse_out array holds #queries
+          chunks of size (2 + N_ROWS_IN_MASK) with the following structure:
+          0..(N_ROWS_IN_MASK-1):   [(x - g^i * z)^(-1) for i in rowsInMask]
+          N_ROWS_IN_MASK:          (x - z^constraintDegree)^-1
+          N_ROWS_IN_MASK+1:        frieval_pointInv.
+    */
+    fun oods_prepare_inverses(ctx: &vector<u256>, batch_inverse_array: &mut vector<u256>) {
+        let eval_coset_offset_ = GENERATOR_VAL;
+        // The array expmods_and_points stores subexpressions that are needed
+        // for the denominators computation.
+        // The array is segmented as follows:
+        //    expmods_and_points[0:13] (.expmods) expmods used during calculations of the points below.
+        //    expmods_and_points[13:111] (.points) points used during the denominators calculation.
+        let expmods_and_points = &mut vector[];
+        {
+            let trace_generator = /*trace_generator*/ *borrow(ctx, MM_TRACE_GENERATOR);
+            let prime = K_MODULUS;
+
+            // Prepare expmods for computations of trace generator powers.
+
+            let tg2 = fmul(trace_generator, trace_generator);
+            let tg3 = fmul(tg2, trace_generator);
+            let tg4 = fmul(tg3, trace_generator);
+            let tg5 = fmul(tg4, trace_generator);
+            let tg7 = fmul(tg4, tg3);
+            let tg12 = fmul(tg7, tg5);
+            let tg13 = fmul(tg12, trace_generator);
+            let tg28 = fmul(tg2, fmul(tg13, tg13));
+            let tg48 = fmul(tg28, fmul(tg13, tg7));
+            let tg216 = fpow(tg12, 18);
+            let tg245 = fmul(trace_generator, fmul(tg216, tg28));
+            let tg320 = fmul(tg216, fmul(tg48, fmul(tg28, tg28)));
+            let tg1010 = fmul(tg2, fmul(tg48, fpow(tg320, 3)));
```

Comment:
```suggestion
            let tg1010 = fmul(tg2, fmul(tg48, fmul(tg320, fmul(tg320, tg320))));
```

Reply:
let tg2 = fmul(trace_generator, trace_generator);
            let tg3 = fmul(tg2, trace_generator);
            let tg4 = fmul(tg3, trace_generator);
            let tg5 = fmul(tg4, trace_generator);
            let tg7 = fmul(tg4, tg3);
            let tg12 = fmul(tg7, tg5);
            let tg13 = fmul(tg12, trace_generator);
            let tg24 = fmul(tg12, tg12);
            let tg28 = fmul(tg24, tg4);
            let tg48 = fmul(tg24, tg24);
            let tg96 = fmul(tg48, tg48);
            let tg192 = fmul(tg96, tg96);
            let tg216 = fmul(tg24, tg192);
            let tg245 = fmul(trace_generator, fmul(tg216, tg28));
            let tg320 = fmul(tg216, fmul(tg48, fmul(tg28, tg28)));
            let tg1010 = fmul(tg2, fmul(tg48, fmul(tg320, fmul(tg320, tg320))));

Reply:
These variables are used for calculations in the rest of the code, so variables such as `tg48` or `tg96` would not be appropriate. I think using `tg216 = fpow(tg12, 18)` doesn't make too much difference.